### ISSUE TITLE
akka-remote not excluded during assembly

### ISSUE THREAD

**Issue by jonbuffington** [spark-jobserver/spark-jobserver#644]
Ran into a class verifier exception yesterday deploying SJS to a new cluster. Tracked it down to a bug/missing sbt feature in the exclude method <https://github.com/sbt/sbt/issues/1518>. `"com.typesafe.akka" %% "akka-cluster" % "2.3.15" exclude("com.typesafe.akka", "akka-remote")` does not exclude akka-remote leading to an old version of netty being assembled into the jar.

My crude work around was <https://github.com/fuseelements/spark-jobserver/commit/1ecee0a3f124f9d86eeb1e07ddc4904b9c34ac0d>. Any suggestions for creating a better commit and opening a pull request?

**Comment by velvia** [comment]
:(  bummer that this hasn’t been fixed in SBT yet.   Yuck.

I can’t think of a much better way, but there is a way to reference the current Scala version, have you tried scalaVersion.value ?

Actually, instead of excluding akka-remote (which is in Spark itself too), what if we excluded Netty instead?

**Comment by jonbuffington** [comment]
@velvia 

`> project akka-app`

Using `com.typesafe.akka" %% "akka-cluster" % "2.3.15" exclude("com.typesafe.akka", "akka-remote")`:
```
> dependencyTree
[warn] Credentials file /Users/jcb/.bintray/.credentials does not exist
[info] Updating {file:/Users/jcb/Projects/spark-jobserver/}akka-app...
[info] Resolving com.googlecode.concurrentlinkedhashmap#concurrentlinkedhashmap-[info] Resolving com.sun.jersey.jersey-test-framework#jersey-test-framework-griz[info] Resolving com.fasterxml.jackson.module#jackson-module-scala_2.10;2.4.4 ..[info] Resolving org.fusesource.jansi#jansi;1.4 ...
[info] Done updating.
[info] spark.jobserver:akka-app_2.10:0.6.2 [S]
[info]   +-com.typesafe.akka:akka-cluster_2.10:2.3.15 [S]
[info]   | +-com.typesafe.akka:akka-remote_2.10:2.3.15 [S]
[info]   |   +-com.google.protobuf:protobuf-java:2.5.0
[info]   |   +-com.typesafe.akka:akka-actor_2.10:2.3.15 [S]
[info]   |   | +-com.typesafe:config:1.2.1 (evicted by: 1.3.0)
[info]   |   | +-com.typesafe:config:1.3.0
[info]   |   | 
[info]   |   +-io.netty:netty:3.8.0.Final
[info]   |   +-org.uncommons.maths:uncommons-maths:1.2.2a
[info]   |   
[info]   +-com.typesafe:config:1.2.1 (evicted by: 1.3.0)
[info]   +-com.typesafe:config:1.3.0
[info]   +-com.yammer.metrics:metrics-core:2.2.0
[info]   | +-org.slf4j:slf4j-api:1.7.2
[info]   | 
[info]   +-io.netty:netty-all:4.0.29.Final
[info]   +-io.spray:spray-caching_2.10:1.3.3 [S]
[info]   | +-com.googlecode.concurrentlinkedhashmap:concurrentlinkedhashmap-l..
[info]   | +-io.spray:spray-util_2.10:1.3.3 [S]
[info]   | 
[info]   +-io.spray:spray-can_2.10:1.3.3 [S]
[info]   | +-io.spray:spray-http_2.10:1.3.3 [S]
[info]   | | +-io.spray:spray-util_2.10:1.3.3 [S]
[info]   | | +-org.parboiled:parboiled-scala_2.10:1.1.7 [S]
[info]   | |   +-org.parboiled:parboiled-core:1.1.7
[info]   | |   
[info]   | +-io.spray:spray-io_2.10:1.3.3 [S]
[info]   | | +-io.spray:spray-http_2.10:1.3.3 [S]
[info]   | | | +-io.spray:spray-util_2.10:1.3.3 [S]
[info]   | | | +-org.parboiled:parboiled-scala_2.10:1.1.7 [S]
[info]   | | |   +-org.parboiled:parboiled-core:1.1.7
[info]   | | |   
[info]   | | +-io.spray:spray-util_2.10:1.3.3 [S]
[info]   | | 
[info]   | +-io.spray:spray-util_2.10:1.3.3 [S]
[info]   | 
[info]   +-io.spray:spray-client_2.10:1.3.3 [S]
[info]   | +-io.spray:spray-can_2.10:1.3.3 [S]
[info]   | | +-io.spray:spray-http_2.10:1.3.3 [S]
[info]   | | | +-io.spray:spray-util_2.10:1.3.3 [S]
[info]   | | | +-org.parboiled:parboiled-scala_2.10:1.1.7 [S]
[info]   | | |   +-org.parboiled:parboiled-core:1.1.7
[info]   | | |   
[info]   | | +-io.spray:spray-io_2.10:1.3.3 [S]
[info]   | | | +-io.spray:spray-http_2.10:1.3.3 [S]
[info]   | | | | +-io.spray:spray-util_2.10:1.3.3 [S]
[info]   | | | | +-org.parboiled:parboiled-scala_2.10:1.1.7 [S]
[info]   | | | |   +-org.parboiled:parboiled-core:1.1.7
[info]   | | | |   
[info]   | | | +-io.spray:spray-util_2.10:1.3.3 [S]
[info]   | | | 
[info]   | | +-io.spray:spray-util_2.10:1.3.3 [S]
[info]   | | 
[info]   | +-io.spray:spray-http_2.10:1.3.3 [S]
[info]   | | +-io.spray:spray-util_2.10:1.3.3 [S]
[info]   | | +-org.parboiled:parboiled-scala_2.10:1.1.7 [S]
[info]   | |   +-org.parboiled:parboiled-core:1.1.7
[info]   | |   
[info]   | +-io.spray:spray-httpx_2.10:1.3.3 [S]
[info]   | | +-io.spray:spray-http_2.10:1.3.3 [S]
[info]   | | | +-io.spray:spray-util_2.10:1.3.3 [S]
[info]   | | | +-org.parboiled:parboiled-scala_2.10:1.1.7 [S]
[info]   | | |   +-org.parboiled:parboiled-core:1.1.7
[info]   | | |   
[info]   | | +-io.spray:spray-util_2.10:1.3.3 [S]
[info]   | | +-org.jvnet.mimepull:mimepull:1.9.5
[info]   | | 
[info]   | +-io.spray:spray-util_2.10:1.3.3 [S]
[info]   | 
[info]   +-io.spray:spray-json_2.10:1.3.2 [S]
[info]   +-io.spray:spray-routing_2.10:1.3.3 [S]
[info]   | +-com.chuusai:shapeless_2.10:1.2.4 [S]
[info]   | +-io.spray:spray-http_2.10:1.3.3 [S]
[info]   | | +-io.spray:spray-util_2.10:1.3.3 [S]
[info]   | | +-org.parboiled:parboiled-scala_2.10:1.1.7 [S]
[info]   | |   +-org.parboiled:parboiled-core:1.1.7
[info]   | |   
[info]   | +-io.spray:spray-httpx_2.10:1.3.3 [S]
[info]   | | +-io.spray:spray-http_2.10:1.3.3 [S]
[info]   | | | +-io.spray:spray-util_2.10:1.3.3 [S]
[info]   | | | +-org.parboiled:parboiled-scala_2.10:1.1.7 [S]
[info]   | | |   +-org.parboiled:parboiled-core:1.1.7
[info]   | | |   
[info]   | | +-io.spray:spray-util_2.10:1.3.3 [S]
[info]   | | +-org.jvnet.mimepull:mimepull:1.9.5
[info]   | | 
[info]   | +-io.spray:spray-util_2.10:1.3.3 [S]
[info]   | 
[info]   +-joda-time:joda-time:2.9.3
[info]   +-org.joda:joda-convert:1.8.1
[info]
```

Using `com.typesafe.akka" %% "akka-cluster" % "2.3.15" exclude("org.jboss.netty", "netty")`:
```
> dependencyTree
[warn] Credentials file /Users/jcb/.bintray/.credentials does not exist
[info] Updating {file:/Users/jcb/Projects/spark-jobserver/}akka-app...
[info] Resolving com.googlecode.concurrentlinkedhashmap#concurrentlinkedhashmap-[info] Resolving com.sun.jersey.jersey-test-framework#jersey-test-framework-griz[info] Resolving com.fasterxml.jackson.module#jackson-module-scala_2.10;2.4.4 ..[info] Resolving org.fusesource.jansi#jansi;1.4 ...
[info] Done updating.
[info] spark.jobserver:akka-app_2.10:0.6.2 [S]
[info]   +-com.typesafe.akka:akka-cluster_2.10:2.3.15 [S]
[info]   | +-com.typesafe.akka:akka-remote_2.10:2.3.15 [S]
[info]   |   +-com.google.protobuf:protobuf-java:2.5.0
[info]   |   +-com.typesafe.akka:akka-actor_2.10:2.3.15 [S]
[info]   |   | +-com.typesafe:config:1.2.1 (evicted by: 1.3.0)
[info]   |   | +-com.typesafe:config:1.3.0
[info]   |   | 
[info]   |   +-org.uncommons.maths:uncommons-maths:1.2.2a
[info]   |   
[info]   +-com.typesafe:config:1.2.1 (evicted by: 1.3.0)
[info]   +-com.typesafe:config:1.3.0
[info]   +-com.yammer.metrics:metrics-core:2.2.0
[info]   | +-org.slf4j:slf4j-api:1.7.2
[info]   | 
[info]   +-io.netty:netty-all:4.0.29.Final
[info]   +-io.spray:spray-caching_2.10:1.3.3 [S]
[info]   | +-com.googlecode.concurrentlinkedhashmap:concurrentlinkedhashmap-l..
[info]   | +-io.spray:spray-util_2.10:1.3.3 [S]
[info]   | 
[info]   +-io.spray:spray-can_2.10:1.3.3 [S]
[info]   | +-io.spray:spray-http_2.10:1.3.3 [S]
[info]   | | +-io.spray:spray-util_2.10:1.3.3 [S]
[info]   | | +-org.parboiled:parboiled-scala_2.10:1.1.7 [S]
[info]   | |   +-org.parboiled:parboiled-core:1.1.7
[info]   | |   
[info]   | +-io.spray:spray-io_2.10:1.3.3 [S]
[info]   | | +-io.spray:spray-http_2.10:1.3.3 [S]
[info]   | | | +-io.spray:spray-util_2.10:1.3.3 [S]
[info]   | | | +-org.parboiled:parboiled-scala_2.10:1.1.7 [S]
[info]   | | |   +-org.parboiled:parboiled-core:1.1.7
[info]   | | |   
[info]   | | +-io.spray:spray-util_2.10:1.3.3 [S]
[info]   | | 
[info]   | +-io.spray:spray-util_2.10:1.3.3 [S]
[info]   | 
[info]   +-io.spray:spray-client_2.10:1.3.3 [S]
[info]   | +-io.spray:spray-can_2.10:1.3.3 [S]
[info]   | | +-io.spray:spray-http_2.10:1.3.3 [S]
[info]   | | | +-io.spray:spray-util_2.10:1.3.3 [S]
[info]   | | | +-org.parboiled:parboiled-scala_2.10:1.1.7 [S]
[info]   | | |   +-org.parboiled:parboiled-core:1.1.7
[info]   | | |   
[info]   | | +-io.spray:spray-io_2.10:1.3.3 [S]
[info]   | | | +-io.spray:spray-http_2.10:1.3.3 [S]
[info]   | | | | +-io.spray:spray-util_2.10:1.3.3 [S]
[info]   | | | | +-org.parboiled:parboiled-scala_2.10:1.1.7 [S]
[info]   | | | |   +-org.parboiled:parboiled-core:1.1.7
[info]   | | | |   
[info]   | | | +-io.spray:spray-util_2.10:1.3.3 [S]
[info]   | | | 
[info]   | | +-io.spray:spray-util_2.10:1.3.3 [S]
[info]   | | 
[info]   | +-io.spray:spray-http_2.10:1.3.3 [S]
[info]   | | +-io.spray:spray-util_2.10:1.3.3 [S]
[info]   | | +-org.parboiled:parboiled-scala_2.10:1.1.7 [S]
[info]   | |   +-org.parboiled:parboiled-core:1.1.7
[info]   | |   
[info]   | +-io.spray:spray-httpx_2.10:1.3.3 [S]
[info]   | | +-io.spray:spray-http_2.10:1.3.3 [S]
[info]   | | | +-io.spray:spray-util_2.10:1.3.3 [S]
[info]   | | | +-org.parboiled:parboiled-scala_2.10:1.1.7 [S]
[info]   | | |   +-org.parboiled:parboiled-core:1.1.7
[info]   | | |   
[info]   | | +-io.spray:spray-util_2.10:1.3.3 [S]
[info]   | | +-org.jvnet.mimepull:mimepull:1.9.5
[info]   | | 
[info]   | +-io.spray:spray-util_2.10:1.3.3 [S]
[info]   | 
[info]   +-io.spray:spray-json_2.10:1.3.2 [S]
[info]   +-io.spray:spray-routing_2.10:1.3.3 [S]
[info]   | +-com.chuusai:shapeless_2.10:1.2.4 [S]
[info]   | +-io.spray:spray-http_2.10:1.3.3 [S]
[info]   | | +-io.spray:spray-util_2.10:1.3.3 [S]
[info]   | | +-org.parboiled:parboiled-scala_2.10:1.1.7 [S]
[info]   | |   +-org.parboiled:parboiled-core:1.1.7
[info]   | |   
[info]   | +-io.spray:spray-httpx_2.10:1.3.3 [S]
[info]   | | +-io.spray:spray-http_2.10:1.3.3 [S]
[info]   | | | +-io.spray:spray-util_2.10:1.3.3 [S]
[info]   | | | +-org.parboiled:parboiled-scala_2.10:1.1.7 [S]
[info]   | | |   +-org.parboiled:parboiled-core:1.1.7
[info]   | | |   
[info]   | | +-io.spray:spray-util_2.10:1.3.3 [S]
[info]   | | +-org.jvnet.mimepull:mimepull:1.9.5
[info]   | | 
[info]   | +-io.spray:spray-util_2.10:1.3.3 [S]
[info]   | 
[info]   +-joda-time:joda-time:2.9.3
[info]   +-org.joda:joda-convert:1.8.1
[info]   
```
By excluding the older netty, the org.jboss.netty classes are not assembled. Should I create a pull request that uses `"com.typesafe.akka" %% "akka-cluster" % "2.3.15" exclude("io.netty", "netty")`?

**Comment by velvia** [comment]
Please go ahead, thanks.  You are confirming that excluding netty by itself works right?

**Comment by jonbuffington** [comment]
@velvia Yes. Excluding the old version of netty (io.netty:netty:3.8.0.Final) by itself works. Creating the pull request now.

**Comment by noorul** [comment]
Closing as this is fixed in #647
